service-base/src/main/java/org/eclipse/hono/service/auth/device/UsernamePasswordCredentials.java

```diff
@@ -114,27 +115,31 @@ public final String getPassword() {
     @Override
     public boolean matchesCredentials(final JsonObject candidateSecret) {
 
+        encoder = PasswordEncoderFactory.createDelegatingPasswordEncoder();
+
         try {
             final String pwdHash = candidateSecret.getString(CredentialsConstants.FIELD_SECRETS_PWD_HASH);
             if (pwdHash == null) {
                 LOG.debug("candidate hashed-password secret does not contain a pwd hash");
                 return false;
             }
 
-            final byte[] hashedPasswordOnRecord = Base64.getDecoder().decode(pwdHash);
+            final String hashFunction = candidateSecret.getString(
+                    CredentialsConstants.FIELD_SECRETS_HASH_FUNCTION,
+                    CredentialsConstants.DEFAULT_HASH_FUNCTION);
+
+            String encodedPassword = "{" + hashFunction + "}";
 
-            byte[] salt = null;
             final String encodedSalt = candidateSecret.getString(CredentialsConstants.FIELD_SECRETS_SALT);
             // the salt is optional so decodedSalt may stay null if salt was not found
             if (encodedSalt != null) {
-                salt = Base64.getDecoder().decode(encodedSalt);
+                final byte[] salt = Base64.getDecoder().decode(encodedSalt);
+                encodedPassword += "{" + new String(salt, StandardCharsets.UTF_8) + "}";
```

Comment:
I am not sure if this works for all hashed passwords that have been created with a byte array that does not represent valid UTF_8 encoded characters. When creating String from such a byte array, you will probably run into an exception or some bytes will be replaced by *default* UTF8 chars. However, this is a migration issue ...